exec/cnex/lib/sqlite.c

```diff
@@ -0,0 +1,407 @@
+
+void sqlite_execOne(TExecutor *exec)
+{
+    Cell *parameters = cell_fromCell(top(exec->stack)); pop(exec->stack);
+    char *sql = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Object *db = check_database(top(exec->stack)); pop(exec->stack);
+
+    sqlite3_stmt *stmt = NULL;
+
+    if (db == NULL) {
+        exec->rtl_raise(exec, "SqliteException.InvalidDatabase", "", BID_ZERO);
+        goto cleanup;
+    }
+
+    Cell *result = cell_createArrayCell(0);
+    BOOL Retval = FALSE;
+
+    int r = sqlite3_prepare_v2(db->ptr, sql, -1, &stmt, NULL);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+        goto cleanup;
+    }
+    for (int i = 0; i < parameters->dictionary->len; i++) {
+        int c = sqlite3_bind_parameter_index(stmt, TCSTR(parameters->dictionary->data[i].key));
+        if (c == 0) {
+            exec->rtl_raise(exec, "SqlException.ParameterName", TCSTR(parameters->dictionary->data[i].key), BID_ZERO);
+            goto cleanup;
+        }
+        r = sqlite3_bind_text(stmt, c, TCSTR(parameters->dictionary->data[i].value->string), -1, SQLITE_TRANSIENT);
+        if (r != SQLITE_OK) {
+            exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+            goto cleanup;
+        }
+    }
+    int columns = sqlite3_column_count(stmt);
+    r = sqlite3_step(stmt);
+    if (r == SQLITE_DONE) {
+        goto done;
+    }
+    if (r == SQLITE_ROW) {
+        for (int i = 0; i < columns; i++) {
+            cell_arrayAppendElementPointer(result, cell_fromCString((const char *)sqlite3_column_text(stmt, i)));
+        }
+    } else {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+        goto cleanup;
+    }
+    Retval = TRUE;
+
+done:
+    push(exec->stack, cell_fromBoolean(Retval));
+    push(exec->stack, result);
+
+cleanup:
+    sqlite3_finalize(stmt);
+    free(sql);
+    cell_freeCell(parameters);
+}
+
+void sqlite_execRaw(TExecutor *exec)
+{
+    char *sql = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Object *db = check_database(top(exec->stack)); pop(exec->stack);
+
+    if (db == NULL) {
+        exec->rtl_raise(exec, "SqliteException.InvalidDatabase", "", BID_ZERO);
+        goto cleanup;
+    }
+
+    Cell *rows = cell_createArrayCell(0);
+    char *errmsg = NULL;
+    int r = sqlite3_exec(db->ptr, sql, callback, rows, &errmsg);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+        goto cleanup;
+    }
+
+    push(exec->stack, rows);
+
+cleanup:
+    free(sql);
+}
+
+void sqlite_close(TExecutor *exec)
+{
+    Object *db = check_database(top(exec->stack)); pop(exec->stack);
+
+    if (db == NULL) {
+        exec->rtl_raise(exec, "SqliteException.InvalidDatabase", "", BID_ZERO);
+        return;
+    }
+
+    int retval = sqlite3_close(db->ptr);
```

Comment:
According to https://sqlite.org/c3ref/close.html, the `sqlite3_close_v2` function would probably be more appropriate here. As a bonus you won't get the `SQLITE_BUSY` return value.